src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const historyArray = [];
+  let changedObject = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      changedObject = {};
+      historyArray.push({ ...changedObject });
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(changedObject, action.extraData);
+      historyArray.push({ ...changedObject });
+    }
+
+    if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => delete changedObject[key]);
+      historyArray.push({ ...changedObject });
+    }
```

Comment:
Consider using a switch statement instead of multiple if statements when checking the `action.type`. This will improve readability and make it easier to manage the different cases.